lib/ansible/config/manager.py

```diff
@@ -143,15 +145,15 @@ def ensure_type(value, value_type, origin=None):
         elif value_type in ('str', 'string'):
             if isinstance(value, (string_types, AnsibleVaultEncryptedUnicode, bool, int, float, complex)):
                 value = to_text(value, errors='surrogate_or_strict')
-                if origin == 'ini':
+                if origin_ftype == 'ini':
```

Comment:
tempted to just fix this and make origin `ini: <file>` the same way we do with `env:` and then instead of `==` use `.startswith('ini')`

Reply:
poc:
```
--- a/lib/ansible/config/manager.py
+++ b/lib/ansible/config/manager.py
@@ -143,7 +143,7 @@ def ensure_type(value, value_type, origin=None):
         elif value_type in ('str', 'string'):
             if isinstance(value, (string_types, AnsibleVaultEncryptedUnicode, bool, int, float, complex)):
                 value = to_text(value, errors='surrogate_or_strict')
-                if origin == 'ini':
+                if origin.startswith('ini: '):
                     value = unquote(value)
             else:
                 errmsg = 'string'
@@ -520,24 +520,32 @@ class ConfigManager(object):
             if self._parsers.get(cfile, None) is None:
                 self._parse_config_file(cfile)
 
+            # attempt to read from config file
             if value is None and cfile is not None:
                 ftype = get_config_type(cfile)
                 if ftype and defs[config].get(ftype):
-                    if ftype == 'ini':
-                        # load from ini config
-                        try:  # FIXME: generalize _loop_entries to allow for files also, most of this code is dupe
-                            for ini_entry in defs[config]['ini']:
-                                temp_value = get_ini_config_value(self._parsers[cfile], ini_entry)
-                                if temp_value is not None:
-                                    value = temp_value
-                                    origin = cfile
-                                    if 'deprecated' in ini_entry:
-                                        self.DEPRECATED.append(('[%s]%s' % (ini_entry['section'], ini_entry['key']), ini_entry['deprecated']))
-                        except Exception as e:
-                            sys.stderr.write("Error while loading ini config %s: %s" % (cfile, to_native(e)))
-                    elif ftype == 'yaml':
-                        # FIXME: implement, also , break down key from defs (. notation???)
-                        origin = cfile
+                    try:
+                        for entry in defs[config][ftype]:
+                            # load from config
+                            if ftype == 'ini':
+                                temp_value = get_ini_config_value(self._parsers[cfile], entry)
+                            elif ftype == 'yaml':
+                                raise AnsibleError('YAML configuration type has not been implemented yet')
+                            else:
+                                raise AnsibleError('Invalid configuration file type: %s' % ftype)
+
+                            if temp_value is not None:
+                                # set value and origin
+                                value = temp_value
+                                origin = '%s: %s' % (ftype, cfile)
+                                if 'deprecated' in entry:
+                                    if ftype == 'ini':
+                                        self.DEPRECATED.append(('[%s]%s' % (entry['section'], entry['key']), entry['deprecated']))
+                                    else:
+                                        raise AnsibleError('Unimplemented file type: %s' % ftype)
+
+                    except Exception as e:
+                        sys.stderr.write("Error while loading ini config %s: %s" % (cfile, to_native(e)))
 
             # set default if we got here w/o a value
             if value is None:
```
i also went ahead and generalized the file code in hopes i'll soon implement yaml config files ...

Reply:
@snipfoo, no, if you want to continue with this PR you can, I just end up 'thinking in code' most of the time and it is easier for me to iterate and/or discuss a PR/branch i publish (something tangible) vs a lengthy explanation.

fyi, i also saw the `None` issue and fixed it but i had no plans of going further, if you want to incorporate the suggested changes here we can move ahead.

Reply:
@bcoca Okay, sure, I'll do that, thanks! I'm not sure how I can attribute the commit to you, though. I can reference your comment here in the commit message, if that's okay with you? or maybe you can amend the commit so as to appear as its author? or perhaps there's a Github feature for doing just that that I don't know about? Sorry, I don't want to overstep here, nor appear like I'm taking credit for an idea of yours.

Reply:
Well, here goes :) I hope that's okay like this. Thanks again!